### ISSUE TITLE
Undefined reference to void Foam 

### ISSUE THREAD

**Issue by SaurabhKavi** [Atizar/RapidCFD-dev#72]
Dear All,

I have cloned the RapidCFD repo and followed the build procedure as usual. My system configuration is 
centos-7, cuda-10.2, nvidia V100 gpu.
steps i did are:-
•	changed architecture to sm_70 in wmake/rules folder.
•	export CUDA_HOME=/usr/local/cuda
•	export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64
•	export PATH=$PATH:$CUDA_HOME/bin

Then i give ./Allwmake but after generation of .so files, i got this error during the generation of solvers-
RapidCFD-dev/platforms/linux64NvccDPOpt/lib/libOpenFOAM.so: undefined reference to `void Foam::AINVPreconditioner::preconditionImpl<false>(Foam::gpuField<double>&, Foam::gpuField<double> const&, unsigned char) const' 
collect2: error: ld returned 1 exit status

This error is coming for all the solvers. Can someone please tell where i am going wrong during the build procedure because I am unable to resolve this. Sorry for asking such basic questions.

**Comment by TonkomoLLC** [comment]
Hello, 

Your error description implied that you started from a fresh install (cloning not pulling from the repo). Please kindly confirm.  

I have compiled commit 4bcd738741ab474dc9a66b6cdf5f1f55308a9326 on a few different machines and have not seen this error.

However, when some of the updates to RapidCFD were posted a couple of months ago, I ran into some compilation errors.  I then tried `wclean all` from the RapidCFD root directory, and the compilation errors vanished.

If this advice is applicable to your situation, I hope you are able to now compile RapidCFD.

However, if you are seeing this error on a fresh install, I do not know what could be causing this error.  I have not seen it, and you may need to search for errors earlier in the compilation process (i.e., the solvers may have compiling errors because the source that comprises libOpenFOAM.so has some compilation errors).

Best regards,

Eric

**Comment by SaurabhKavi** [comment]
Hi Eric,
Yes I started from a fresh install (cloned) and followed the steps I mentioned above. As you suggested I did wclean all and started again, but the same error “undefined reference to `void Foam::AINVPreconditioner::preconditionImpl<false>(Foam::gpuField<double>&, Foam::gpuField<double> const&, unsigned char) const'” occurs.
Also I am getting this warning during compilation “catching polymorphic type ‘class std::bad_cast’ by value”, but I think we can ignore this. Also i had one question do we need to install OpenFOAM before building RapidCFD ?
Sorry for being late. I am attaching my error log file if needed but it's quite large.
[error_log.txt](https://github.com/Atizar/RapidCFD-dev/files/4465037/error_log.txt)

Thanks.

**Comment by TonkomoLLC** [comment]
Hi,
To answer your question: no, you do not need to install OpenFOAM before building RapidCFD. 

Hmm. I do not know what the solution is, off-hand.  I have not seen these compilation errors before.  I have tested with Ubuntu 16.04 and 18.04, gcc 7.4.0, and CUDA 8 & CUDA 10.2.  

Idea: You can try installing an old commit -  commit f3775ac96129bfee68655e11e63ff5d62bccb4b9 and see if that works. There were a lot of good improvements that were done since this commit, but this commit worked for me across a number of machines/configurations.

I do not have a specific idea beyond this because I have not yet encountered the errors you are describing. 

Best regards,

Eric

**Comment by SaurabhKavi** [comment]
Hi Eric,
Thanks for the help. I am now able to successfully compile RapidCFD on my system. The main problem was with my cuda installation. After the reinstallation, there were no more build errors.
Are there any simple examples by which I can verify my GPU usage. Need some help from you on this. I see there is a RapidCFD Tests repo hosted by you and was checking how to use those testcases with my RapidCFD. But was unable to do so. Little insights on this would be very helpful.

Best regards
Saurabh.

**Comment by TonkomoLLC** [comment]
Glad you found the problem and can compile RapidCFD.  On the test cases, firstly, you need OpenFOAM installed in order to pre-process and post-process (e.g., blockMesh, setFields if required, and paraFoam).  I guess my answer above should be clarified: No, you do not need OpenFOAM previously installed to **build** RapidCFD, but you need OpenFOAM installed to **use** RapidCFD.

The easiest test to run is `cavity` found [here](https://github.com/TonkomoLLC/RapidCFD-Tests/tree/master/cavity).  Since the grid is already created, you can just run `icoFoam` from the case directory.  To be clear, this test case is so small that it will not be as fast as CPU-based OpenFOAM, but it will quickly indicate if your RapidCFD setup is working.  

If you have a specific question or request please let me know.

Thanks,

Eric

**Comment by TonkomoLLC** [comment]
To the best of my knowledge, solvers which depend on other libraries which have not been ported to CUDA are not included.  So, for example, reactingFoam is not included because the whole lot of the chemistry capability is not ported to CUDA. I do not know all the underlying reasons as to why some solvers and/or underlying models are included and others are not, but I suspect that there are constraints both of time and what's impactful with GPGPU programming. 

Others are welcome to correct me, but I think that linear-type problems (Ax=b) work well on the GPU, but other types of problems may not.  Data transfer between the CPU and GPU can be expensive.  For this reason, and as an example, I think chemical source terms (rate ~ exp(-activation_energy/R/T)) may not work so well on GPU CFD. It is for this reason I believe reactingFoam is not included.  However, I want to underscore this is my impression as to why reactingFoam is missing, and I have not seen an "official reason." 

I hope this is helpful (and accurate).  I welcome others to chime in.

Cheers,
Eric

**Comment by SaurabhKavi** [comment]
Hi Eric,
Thanks for being so helpful. I have one question, can we build RapidCFD with Clang compiler as well. Would it work with latest cuda installation or do I need to have older versions of cuda for it. 
Best Regards
Saurabh

**Comment by TonkomoLLC** [comment]
Hello, Saurabh,

Re: Clang, see the discussion at the bottom of issue #70 

I never got Clang to compile RapidCFD properly, but maybe I did not do something right. The author of RapidCFD is able to compile with Clang.  As we discussed in that referenced issue, I think you do indeed need to use an older version of CUDA.  I think I was using Clang 9 and CUDA 8 in my trials, but now I see Clang 10 is out and it is compatible with CUDA up to [v10.1](https://llvm.org/docs/CompileCudaWithLLVM.html).  I have not attempted this new combination of Clang and CUDA.

Good luck -

Cheers,
Eric

**Comment by SaurabhKavi** [comment]
Hi Eric.
I tried compiling with clang but was not successful in my attempt.
So I continued with nvcc for now. Can you give some help on why dambreakrapidCFD test is failing for me. Below is the error log:-
--> FOAM FATAL IO ERROR:
keyword nu is undefined in dictionary "/saurabh/RapidCFD/RapidCFD-Tests/damBreak-rapidCFD/constant/transportProperties"
file: /saurabh/RapidCFD/RapidCFD-Tests/damBreak-rapidCFD/constant/transportProperties from line 18 to line 64.
From function dictionary::lookupEntry(const word&, bool, bool) const
in file db/dictionary/dictionary.C at line 437.

My system specification are:- 
Ubuntu-18.04, Cuda-10.2, Tesla v100 32gb single gpu.
Is it failing because I am running it on single gpu?

Thanks
Saurabh

**Comment by SaurabhKavi** [comment]
Hi Eric.

I hope you are doing good. I need your guidance in running some OpenFoam examples with RapidCFD solvers. What changes do I need to do in the fvSolution file so that RapidCFD solvers can run. Or if changes are required in some more files as well.

Best Regards.
Saurabh

**Comment by TonkomoLLC** [comment]
Dear Saurabh,

I am not sure if you found this repo yet, but I placed some examples [here](https://github.com/TonkomoLLC/RapidCFD-Tests).  I hope that these examples can help you get started. 

In general, though, not all of the solvers, smoothers and preconditioners available in OpenFOAM are available in RapidCFD.  The good news is that for a lot of circumstances, the numerical tools available are enough to solve a wide range of cases. Moreover, if you select an option in fvSolution that is not encoded in RapidCFD, the runtime selection system will let you know which features are indeed available to select from. You can also hunt through the code at $WM_PROJECT_DIR/src/OpenFOAM/matrices, e.g., [here](https://github.com/Atizar/RapidCFD-dev/tree/master/src/OpenFOAM/matrices/lduMatrix/solvers) and other similar subdirectories in this part of RapidCFD, and see what's available. 

Beyond system/fvSolution, be sure that your case is setup for OpenFOAM 2.3.x, including all of the files in the `constant` directory.


Does this answer your question? 

Best regards,

Eric

**Comment by SaurabhKavi** [comment]
Thanks for the quick informative reply Eric. Yes this answer my question completely, I will look into the points you mentioned and hopefully will be able to run more examples. I will come back here if stuck somewhere.
Thanks for being so helpful

Cheers,
Saurabh

**Comment by SaurabhKavi** [comment]
Hi @TonkomoLLC 
Hope you are doing good. I have a small question regarding the scalability of RapidCFFD. Does it supports Multi Node-Multi GPU ?

Best Regards
Saurabh

**Comment by TonkomoLLC** [comment]
Hello, Saurabh,

Yes, multi-node multi-GPU is possible. See #24 . 

Best regards,

Eric

**Comment by dmaj** [comment]
I ran into the same problem when trying to compile the project.
nvcc: V10.1.243
gcc: 9.3.0 (also 8.xxx)
ubuntu: 20.04

I also get the message that references cannot be resolved: 
"undefined reference to void Foam::AINVPreconditioner::preconditionImpl".

There seems to be a problem with the templates here.

I have rewritten the code in the relevant places and thereby eliminated the error. With the modification I at least got "interFoam" to work. But there seems to be a similar problem with "snappyHexMesh".

I understood what was happening, but unfortunately not yet why it was happening. Since the modification of the two files led to success, I'm probably not completely wrong. I will not invest any more time, but maybe it helps someone who encounters this problem.

From template to function.
preconditionImpl<true>(wA, rA, cmpt); ==> preconditionImpl(true, wA, rA, cmpt);

[AINVPreconditioner.C.txt](https://github.com/Atizar/RapidCFD-dev/files/6643655/AINVPreconditioner.C.txt)
[AINVPreconditioner.H.txt](https://github.com/Atizar/RapidCFD-dev/files/6643657/AINVPreconditioner.H.txt)

**Comment by TonkomoLLC** [comment]
Hi, dmaj,

Thanks for this workaround. Unfortunately, I do not know the implications of your fix, if any. Presently I cannot reproduce the AINVPreconditioner  compilation problem on Ubuntu 20.04, gcc 9.3.0, nvcc 11.1.105.

`snappyHexMesh` is not included with RapidCFD, so I do not follow your comment about the mesher having a compilation problem with AINVPreconditioner.

Best regards,
Eric

**Comment by vdalessa** [comment]
Hi dmaj,
I have tested your patch on V100 with CUDA 10.2 and gnu/8.3.0 .
I have obtained the same results in comparision with OF2.3.0 on standard cavity/icofoam
benchmark.
Valerio

**Comment by Dcn303** [comment]
> 
Hi TonkomoLLC 
where can i find this wclean all , i did not find it in the RapidCFD root directory , but i found the wcleanAll  in the directory
RapidCFD/RapidCFD-dev/wmake 
i.e. in the wmake folder , please correct me 
Thanks 





> Hello,
> 
> Your error description implied that you started from a fresh install (cloning not pulling from the repo). Please kindly confirm.
> 
> I have compiled commit [4bcd738](https://github.com/Atizar/RapidCFD-dev/commit/4bcd738741ab474dc9a66b6cdf5f1f55308a9326) on a few different machines and have not seen this error.
> 
> However, when some of the updates to RapidCFD were posted a couple of months ago, I ran into some compilation errors. I then tried `wclean all` from the RapidCFD root directory, and the compilation errors vanished.
> 
> If this advice is applicable to your situation, I hope you are able to now compile RapidCFD.
> 
> However, if you are seeing this error on a fresh install, I do not know what could be causing this error. I have not seen it, and you may need to search for errors earlier in the compilation process (i.e., the solvers may have compiling errors because the source that comprises libOpenFOAM.so has some compilation errors).
> 
> Best regards,
> 
> Eric

**Comment by TonkomoLLC** [comment]
Hello, 

The `wmake` and `wclean` scripts are the same as those found in CPU versions of OpenFOAM. Once you source the `bashrc` file in /opt/RapidCFD-dev/etc/bashrc  (e.g., `source /opt/RapidCFD-dev/etc/bashrc` , replacing /opt/RapidCFD-dev with your install location) the `wmake` and `wclean` scripts should run from the console (the referenced bashrc script will add paths to your environment to the locations where these scripts are located). 

I hope this answers your question.

**Comment by jasonleong84** [comment]
> Hi Eric, Thanks for the help. I am now able to successfully compile RapidCFD on my system. The main problem was with my cuda installation. After the reinstallation, there were no more build errors. Are there any simple examples by which I can verify my GPU usage. Need some help from you on this. I see there is a RapidCFD Tests repo hosted by you and was checking how to use those testcases with my RapidCFD. But was unable to do so. Little insights on this would be very helpful.
> 
> Best regards Saurabh.

@SaurabhKavi I hope you are still following this thread. I am encountering the same errors when compiling RapidCFD; it compiles the libraries but not the solvers. How did you solve this issue? Thanks -Jason

**Comment by kayog96** [comment]
@jasonleong84. Sorry I lost access to my other account. 
About the issue which i faced long time back worked out after i did a fresh install of my cuda libs. You can give it a try. Hope it works out for you as well.

**Comment by jasonleong84** [comment]
> I ran into the same problem when trying to compile the project. nvcc: V10.1.243 gcc: 9.3.0 (also 8.xxx) ubuntu: 20.04
> 
> I also get the message that references cannot be resolved: "undefined reference to void Foam::AINVPreconditioner::preconditionImpl".
> 
> There seems to be a problem with the templates here.
> 
> I have rewritten the code in the relevant places and thereby eliminated the error. With the modification I at least got "interFoam" to work. But there seems to be a similar problem with "snappyHexMesh".
> 
> I understood what was happening, but unfortunately not yet why it was happening. Since the modification of the two files led to success, I'm probably not completely wrong. I will not invest any more time, but maybe it helps someone who encounters this problem.
> 
> From template to function. preconditionImpl<true>(wA, rA, cmpt); ==> preconditionImpl(true, wA, rA, cmpt);
> 
> [AINVPreconditioner.C.txt](https://github.com/Atizar/RapidCFD-dev/files/6643655/AINVPreconditioner.C.txt) [AINVPreconditioner.H.txt](https://github.com/Atizar/RapidCFD-dev/files/6643657/AINVPreconditioner.H.txt)

Unfortunately, a fresh install of CUDA did not solve my problem but going through this discussion was helpful. By following the suggestions from @dmaj, I was able to compile laplacianFoam. I'll report back if I it works for the rest of the solvers. For reference, my system:

Red Hat Enterprise Linux 8.7 (Ootpa), CUDA Toolkit 11.0, GCC 8.5.0, 2 x V100

**Comment by jasonleong84** [comment]
Happy to report that the modification was successful. I have tested cavity and interFoam test cases posted here:

https://github.com/TonkomoLLC/RapidCFD-Tests

@TonkomoLLC, for dambreak case, I am able to run single GPU case but mpi case is not able to run:

using the following command:

```
mpirun -np 2 interFoam -parallel -devices "(0 1)"

--------------------------------------------------------------------------
It looks like opal_init failed for some reason; your parallel process is
likely to abort.  There are many reasons that a parallel process can
fail during opal_init; some of which are due to configuration or
environment problems.  This failure appears to be an internal failure;
here's some additional information (which may only be relevant to an
Open MPI developer):

  opal_shmem_base_select failed
  --> Returned value -1 instead of OPAL_SUCCESS
--------------------------------------------------------------------------
```

I am currently using the Thirdparty-dev folder you shared with openmpi-2.1.1. @TonkomoLLC would you have any advice here?

**Comment by TonkomoLLC** [comment]
Hi, congratulations on getting RapidCFD working!  As to your mpi issue, I have not come across this one before. The error you report has been discussed elsewhere, like [CFD-Online](https://www.cfd-online.com/Forums/openfoam-solving/117893-problem-running-openfoam-2-2-x-parallel-centos-5-a.html).  Not sure if this link will be helpful, but I hope it is.

Also please see #57: The magic to make a ThirdParty directory work with RapidCFD/GPU's is a small addition to the [Allwmake](https://github.com/TonkomoLLC/ThirdParty-dev/blob/master/Allwmake#L87) file:

```
	# Add CUDA
	configOpt="--with-cuda"
```

Also, in the event you compiled RapidCFD without a ThirdParty directory, then added a ThirdParty directory later, maybe you need to `wclean` then `wmake` the `RapidCFD-dev/src/Pstream` directory. Finally, please be sure that the ThirdParty directory was compiled as usual with ./Allwmake from the ThirdParty-dev directory. These points may be obvious and you probably already took care of these items, but just in case I thought I would mention these ideas.

Good luck!

**Comment by jasonleong84** [comment]
@TonkomoLLC Thanks for the tips. I solved by compiling OpenMPI-1.8.4 in the Thirdparty-dev here: https://sourceforge.net/projects/rapidcfd/files/ after changing some of the flags  for the OpenMPIconfigure portion in the Allwmake file so that:

` --with-cuda=$CUDA_HOME`

I was able to run a dual GPU case for the damBreak Interfoam case you shared.
